projects/l2_pooling/noise_tolerance.py

```diff
@@ -0,0 +1,809 @@
+from htmresearch.frameworks.layers.sensor_placement import greedySensorPositions
+
+L4_CELL_COUNT = 8*1024
+
+
+def createRandomObjectDescriptions(numObjects,
+                                   numLocationsPerObject,
+                                   featurePool=("A", "B", "C")):
+  """
+  Returns {"Object 1": [(0, "C"), (1, "B"), (2, "C"), ...],
+           "Object 2": [(0, "C"), (1, "A"), (2, "B"), ...]}
+  """
+  return dict(("Object %d" % i,
+               zip(xrange(numLocationsPerObject),
+                   [random.choice(featurePool)
+                    for _ in xrange(numLocationsPerObject)]))
+              for i in xrange(1, numObjects + 1))
+
+
+def noisy(pattern, noiseLevel, totalNumCells):
+  n = int(noiseLevel * 40)
+
+  noised = set(pattern)
+
+  noised.difference_update(random.sample(noised, n))
+
+  for _ in xrange(n):
+    while True:
+      v = random.randint(0, totalNumCells - 1)
+      if v not in pattern and v not in noised:
+        noised.add(v)
+        break
+
+  return noised
+
+
+def doExperiment(numColumns, l2Overrides, objectDescriptions, noiseFn,
+                 numInitialTraversals):
+
+  # For each column, keep a mapping from feature-location names to their SDRs
+  layer4sdr = lambda : set(random.sample(xrange(L4_CELL_COUNT), 40))
+  featureLocationSDRs = [defaultdict(layer4sdr) for _ in xrange(numColumns)]
+
+  poolers = [ColumnPooler(inputWidth=L4_CELL_COUNT,
+                          lateralInputWidths=[4096]*(numColumns-1),
+                          **l2Overrides)
+             for _ in xrange(numColumns)]
+
+  # Learn the objects
+  objectL2Representations = {}
+  for objectName, featureLocations in  objectDescriptions.iteritems():
+    for featureLocationName in featureLocations:
+      for _ in xrange(10):
+        allLateralInputs = [pooler.getActiveCells() for pooler in poolers]
+        for columnNumber, pooler in enumerate(poolers):
+          feedforwardInput = featureLocationSDRs[columnNumber][featureLocationName]
+          lateralInputs = [lateralInput
+                           for i, lateralInput in enumerate(allLateralInputs)
+                           if i != columnNumber]
+          pooler.compute(feedforwardInput, lateralInputs, learn=True)
+    objectL2Representations[objectName] = [set(pooler.getActiveCells())
+                                           for pooler in poolers]
+    for pooler in poolers:
+      pooler.reset()
+
+  results = []
+
+  # Try to infer the objects
+  for objectName, featureLocations in objectDescriptions.iteritems():
+    for pooler in poolers:
+      pooler.reset()
+
+    sensorPositionsIterator = greedySensorPositions(numColumns, len(featureLocations))
+
+    # Touch each location at least numInitialTouches times, and then touch it
+    # once more, testing it.
+    numTouchesPerTraversal = len(featureLocations) / float(numColumns)
+    numInitialTouches = int(math.ceil(numInitialTraversals * numTouchesPerTraversal))
```

Comment:
I'm adding some documentation. It traverses the object with the sensors some number of times, then it traverses the object once more. A "traversal" consists of touching every point on the object ~once. Not once per sensor -- that would give an unfair advantage to larger numbers of sensors. Just once, globally. Hence the need to distinguish between a "traversal count" and a "touch count".